internal/process/manager.go

```diff
@@ -0,0 +1,448 @@
+limitations under the License.
+*/
+
+package process
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"net"
+	"os"
+	"path"
+	"reflect"
+	"strings"
+	"sync"
+
+	"github.com/cilium/ebpf"
+	"github.com/cilium/ebpf/link"
+	"github.com/cilium/ebpf/perf"
+	"github.com/containernetworking/plugins/pkg/ns"
+	"github.com/shirou/gopsutil/process"
+	log "github.com/sirupsen/logrus"
+
+	"github.com/merbridge/merbridge/config"
+	"github.com/merbridge/merbridge/internal/ebpfs"
+	"github.com/merbridge/merbridge/pkg/linux"
+)
+
+const (
+	AMBIENT_MESH_MODE_FLAG = (1 << 0) // nolint: stylecheck
+	ZTUNNEL_FLAG           = (1 << 2) // nolint: stylecheck
+	ZTUNNEL_IP_KEY         = 0x1      // nolint: stylecheck
+)
+
+type cgroupInfo struct {
+	// sync from bpf/headers/helpers.h cgroup_info
+	/*
+			__u64 id;
+		    __u32 is_in_mesh;
+		    __u32 cgroup_ip[4];
+		    // We can't specify which ports are listened to here, so we open up a flags,
+		    // user-defined. E.g, for those who wish to determine if port 15001 is
+		    // listened to, we can customize a flag, `IS_LISTEN_15001 = 1 << 2`, which
+		    // we can subsequently detect by `flags & IS_LISTEN_15001`.
+		    __u16 flags;
+		    // detected_flags is used to determine if this operation has ever been
+		    // performed. if `flags & IS_LISTEN_15001` is false but `detected_flags &
+		    // IS_LISTEN_15001` is true, that means real true, we do not need recheck.
+		    // but if `detected_flags & IS_LISTEN_15001` is false, that probably means
+		    // we haven't tested it and need to retest it.
+		    __u16 detected_flags;
+	*/
+	ID            uint64
+	IsInMesh      uint32
+	CgroupIP      [4]uint32
+	Flags         uint16
+	DetectedFlags uint16
+}
+
+type ProcessManager interface {
+	Run(stop chan struct{}) error
+	OnPodStatusChanged(ip string, isInMesh bool, isAmbient bool, isZtunnel bool) error
+}
+
+type podInfo struct {
+	ip        string
+	isInMesh  bool
+	isAmbient bool
+	isZtunnel bool
+}
+
+type processManager struct {
+	lock            sync.RWMutex
+	cgroupMountPath string
+	pidCgroupMap    map[uint32]uint64
+	cgroupIPMap     map[uint64]string
+	ipCgroupsMap    map[string]map[uint64]bool
+	podIPModeMap    map[string]podInfo
+}
+
+func NewProcessManager(cgroupMountPath string) (ProcessManager, error) {
+	if cgroupMountPath == "" {
+		p, err := linux.GetCgroup2MountPath(config.HostProc)
+		if err != nil {
+			return nil, err
+		}
+		if p == "" {
+			// try cgroupv1
+			p, err = linux.GetCgroupSystemdMountPath(config.HostProc)
+			if err != nil {
+				return nil, err
+			}
+			if p == "" {
+				return nil, fmt.Errorf("can not fetch cgroup mount path")
+			}
+		}
+		cgroupMountPath = p
+	}
+	return &processManager{
+		cgroupMountPath: cgroupMountPath,
+		pidCgroupMap:    make(map[uint32]uint64),
+		cgroupIPMap:     make(map[uint64]string),
+		ipCgroupsMap:    make(map[string]map[uint64]bool),
+		podIPModeMap:    make(map[string]podInfo),
```

Comment:
Why maintain these maps?

Reply:
These are mainly used to cache the relationship between PodIP, PID, and Cgroup for the control plane, so there is no need to query from the eBPF map every time.